core/src/main/scala/kafka/server/ConfigAdminManager.scala

```diff
@@ -0,0 +1,518 @@
+      // we should probably only change this in a KIP since it has compatibility implications.
+      case e: Throwable => throw new InvalidRequestException(e.getMessage)
+    }
+ }
+
+  /**
+   * Preprocess a legacy configuration operation on the broker.
+   *
+   * @param request     The request data.
+   *
+   * @return
+   */
+  def preprocess(
+    request: AlterConfigsRequestData,
+  ): util.IdentityHashMap[LAlterConfigsResource, ApiError] = {
+    val results = new util.IdentityHashMap[LAlterConfigsResource, ApiError]()
+    val resourceIds = new util.HashMap[(Byte, String), LAlterConfigsResource]
+    request.resources().forEach(resource => {
+      val preexisting = resourceIds.put((resource.resourceType(), resource.resourceName()), resource)
+      if (preexisting != null) {
+        Seq(preexisting, resource).foreach(
+          r => results.put(r, new ApiError(INVALID_REQUEST, "Each resource must appear at most once.")))
+      }
+    })
+    request.resources().forEach(resource => {
+      if (!results.containsKey(resource)) {
+        val resourceType = ConfigResource.Type.forId(resource.resourceType())
+        val configResource = new ConfigResource(resourceType, resource.resourceName())
+        try {
+          if (containsDuplicates(resource.configs().asScala.map(_.name()))) {
+            throw new InvalidRequestException("Error due to duplicate config keys")
+          }
+          val nullUpdates = new util.ArrayList[String]()
+          resource.configs().forEach { config =>
+            if (config.value() == null) {
+              nullUpdates.add(config.name())
+            }
+          }
+          if (!nullUpdates.isEmpty()) {
+            throw new InvalidRequestException("Null value not supported for : " +
+              String.join(", ", nullUpdates))
+          }
+          resourceType match {
+            case BROKER =>
+              if (!configResource.name().isEmpty) {
+                validateResourceNameIsCurrentNodeId(resource.resourceName())
+              }
+              validateBrokerConfigChange(resource, configResource)
+            case TOPIC =>
+            // Nothing to do.
+            case _ =>
+              // Since legacy AlterConfigs does not support BROKER_LOGGER, any attempt to use it
+              // gets caught by this clause.
+              throw new InvalidRequestException(s"Unknown resource type ${resource.resourceType().toInt}")
+          }
+        } catch {
+          case t: Throwable => {
+            val err = ApiError.fromThrowable(t)
+            info(s"Error preprocessing alterConfigs request on ${configResource}: ${err}")
+            results.put(resource, err)
+          }
+        }
+      }
+    })
+    results
+  }
+
+  def validateBrokerConfigChange(
+    resource: LAlterConfigsResource,
+    configResource: ConfigResource
+  ): Unit = {
+    val props = new Properties()
+    resource.configs().forEach {
+      config => props.put(config.name(), config.value())
```

Comment:
I recommend using `setProperty` as it statically checks that both the key and value are `String`. We have had issues where `Properties` objects contained keys and values that are not `String`.

Reply:
In this case, we do know that the type is `String`. But I agree, in general it is good practice to avoid `Properties#put`. I will change it to `Properties#setProperty`.